### PR TITLE
feat: add var to block custom component execution

### PR DESCRIPTION
**Pull request by jordanrfrazier**
Add an environment variable (`LANGFLOW_ALLOW_CUSTOM_COMPONENTS`) that allows users to run Langflow _without_ allowing execution of custom code. 

Blocks all FE-related custom component creation and editing of components. 

Will still allow you to upload flows with custom components, but will either prompt users with a notification to update, or a notification that it cannot run. 

```
LANGFLOW_ALLOW_CUSTOM_COMPONENTS

    """If set to False, blocks execution of components whose code does not match a known
    server template. The server validates node code against its component template cache;
    when the cache is not yet loaded (e.g., during startup), all flow execution is blocked
    as a safety measure. Set via LANGFLOW_ALLOW_CUSTOM_COMPONENTS environment variable."""
```

Allows users to specify a list of allowed components via the existing `LANGFLOW_COMPONENTS_PATH`.

## Summary by CodeRabbit

* **New Features**
  * Added admin-configurable setting to enable or disable custom component execution in flows.
  * Enhanced component update interface with improved messaging and conditional controls based on custom component permissions.

* **Bug Fixes**
  * Improved error messages with detailed error information extraction across flow endpoints.

* **Chores**
  * Enhanced logging for template validation and component processing debugging.

### CODE REVIEW

**Comment by Cristhianzl** [comment]
## Scoring Summary

| Category | Status | Details |
|----------|--------|---------|
| CRITICAL: Security & PII | PASS | No PII in logs, no hardcoded secrets |
| CRITICAL: DRY Principle | ISSUES FOUND | 3 DRY violations identified |
| CRITICAL: File Structure Limits | PASS | All new/modified files within limits |
| IMPORTANT: Architecture | PASS | Clean separation, fail-closed design |
| IMPORTANT: Code Quality | MINOR ISSUES | Some `any` types in TS, minor magic numbers |
| IMPORTANT: Error Handling | PASS (with notes) | Consistent error handling, some silent returns |
| RECOMMENDED: Observability | PASS | Structured logging at key points |
| RECOMMENDED: Comments | PASS | Docstrings explain "why", no dead code |
| TESTING | PASS (with notes) | Thorough coverage, some gaps noted |

---

## CRITICAL: Security & PII

### PII in Logs — ZERO TOLERANCE

- [x] **No email addresses in logs**
- [x] **No user names in logs**
- [x] **No phone numbers in logs**
- [x] **No PII in webhook messages**
- [x] **No `print()` / `console.log()` with user data**
- [x] **No PII in any logging call**

All log messages use component display names, node IDs, and flow IDs — these are server-side operational data, not PII. Example from `flow_validation.py`:

```python
logger.warning("Blocking flow execution: custom components detected: %s", blocked_labels)
```

**Verdict: PASS**

### General Security

- [x] All user inputs validated — hash verification is server-side, client `edited` flag is NOT trusted
- [x] No secrets/credentials in code
- [x] No internal details exposed to end users — error messages are generic
- [x] Fail-closed design: when hash cache is unavailable (server starting), execution is blocked entirely
- [x] No hardcoded API keys, tokens, or passwords

**Positive highlight:** The security design is excellent. The PR explicitly documents (via code comments and tests) that the client-side `edited` flag is not trusted for security decisions — only server-side hash comparison is used.

**Verdict: PASS**

---

## CRITICAL: DRY Principle

### Violation 1 — Backend endpoint validation pattern (MODERATE)

**Severity:** MODERATE
**Location:** `chat.py` (~5 occurrences), `endpoints.py` (~3 occurrences), `mcp_utils.py`, `openai_responses.py`, `projects.py`, `workflow.py`

The following pattern is repeated ~8 times across endpoint files:

```python
settings_service = get_settings_service()
allow_custom = settings_service.settings.allow_custom_components
if not allow_custom:
    hash_dict = component_cache.type_to_current_hash
    try:
        check_flow_and_raise(flow_data, hash_dict, allow_custom)
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
```

**Recommendation:** Extract to a FastAPI dependency or shared helper:

```python
async def validate_custom_components(flow_data: dict) -> None:
    """FastAPI dependency for custom component validation."""
    settings = get_settings_service().settings
    if not settings.allow_custom_components:
        hash_dict = component_cache.type_to_current_hash
        try:
            check_flow_and_raise(flow_data, hash_dict, settings.allow_custom_components)
        except ValueError as exc:
            raise HTTPException(status_code=400, detail=str(exc)) from exc
```

### Violation 2 — Frontend guard logic duplication (SIGNIFICANT)

**Severity:** SIGNIFICANT
**Location:** `use-post-template-value.ts` and `use-refresh-model-inputs.ts`

The outdated-component guard + 403 suppression logic is duplicated nearly identically:

```typescript
// Duplicated in BOTH files:
const nodeId = data.id;
const componentsToUpdate = useFlowStore.getState().componentsToUpdate;
const matchEntry = componentsToUpdate.find(/* same logic */);
if (matchEntry) {
  return undefined; // silently skip
}
// ... and the 403 catch block with custom component detail check
```

**Recommendation:** Extract to a shared utility, e.g.:

```typescript
// utils/customComponentGuards.ts
export function isNodeOutdated(nodeId: string): boolean { ... }
export function isCustomComponentBlockError(error: unknown): boolean { ... }
```

### Violation 3 — Legacy type aliases defined in 3 places (MODERATE)

**Severity:** MODERATE
**Location:**
1. `src/backend/base/langflow/initial_setup/setup.py` — `_LEGACY_TYPE_ALIASES = {"Prompt": "Prompt Template"}`
2. `src/lfx/src/lfx/interface/components.py` — same dict
3. `src/frontend/src/utils/reactflowUtils.ts` — `LEGACY_TYPE_ALIASES`

All three have `SYNC` comments pointing to each other, which is helpful but doesn't prevent drift. Adding a new alias requires updating all three locations manually.

**Recommendation:** For the backend, define the canonical dict in one place and import from there. The frontend/backend split makes full deduplication difficult, but the backend-side duplication can be resolved.

**Verdict: ISSUES FOUND — 3 violations, none are blockers but should be addressed**

---

## CRITICAL: File Structure Limits

| File | Lines | Functions | Classes | Status |
|------|-------|-----------|---------|--------|
| `flow_validation.py` (NEW) | ~146 | 4 (all validation-related) | 0 | PASS |
| `test_flow_validation.py` (NEW) | ~487 | ~25 (tests) | 4 test classes | PASS (test files exempt) |
| `test_initial_setup.py` (NEW) | ~277 | ~12 (tests) | 2 test classes | PASS (test files exempt) |

- [x] No file exceeds ~500 lines
- [x] No file has more than 5 functions with DIFFERENT responsibilities
- [x] No mixed responsibility prefixes in same file
- [x] Single responsibility per new file

**Verdict: PASS**

---

## IMPORTANT: Architecture & Structure

### Single Responsibility Principle

- [x] `flow_validation.py` — validates flow components against hash index (single responsibility)
- [x] `ComponentCache` additions — builds and stores hash lookups (coherent with existing cache responsibility)
- [x] `settings/base.py` — only adds a new setting field
- [x] Frontend store additions — `pendingNodeUpdates` coordinates build timing (single purpose)

### Layer Separation

- [x] Validation logic is in `api/utils/flow_validation.py` (utility layer)
- [x] Endpoints call validation and convert errors to HTTP responses (handler layer)
- [x] Hash computation is in `lfx/interface/components.py` (service layer)
- [x] No business logic in controllers beyond delegation

### Security Architecture — Positive Highlights

1. **Server-side hash verification:** The `edited` flag from the client is explicitly NOT trusted. The server computes SHA256 hashes of component code and compares against known templates.
2. **Fail-closed design:** When `type_to_current_hash` is `None` (server starting), all execution is blocked.
3. **Recursive validation:** Nested nodes (group nodes) are recursively checked.
4. **All execution paths covered:** chat, endpoints (run/webhook/experimental), MCP, OpenAI Responses, V2 workflow, custom component create/update.

**Verdict: PASS**

---

## IMPORTANT: Code Quality

### TypeScript `any` Types

| Location | Usage | Severity |
|----------|-------|----------|
| `deploy-dropdown.tsx` | `onError: (e: any)` | Low — error handler |
| `use-post-template-value.ts` | `catch (e: any)` | Low — HTTP error |
| `use-save-flow.ts` | `catch (e: any)` | Low — HTTP error |

These are all in error catch blocks where the error type from HTTP libraries is genuinely uncertain. Pragmatic but could use a typed error interface for consistency.

### Magic Numbers

| Location | Value | Context |
|----------|-------|---------|
| `flow_validation.py` | `[:12]` | SHA256 hash prefix length — documented in docstring |
| `flowStore.ts` | `10_000` | Wait timeout for pending updates — documented in function signature |

Both are documented and acceptable.

### Naming

- [x] Functions use clear verbs: `check_flow_and_raise`, `code_hash_matches_any_template`, `_get_invalid_components`
- [x] Boolean state: `allowCustomComponents`, `isRequired`, `dismissed`
- [x] No vague names like `processData` or `handleStuff`

**Verdict: MINOR ISSUES — `any` types in error handlers only**

---

## IMPORTANT: Error Handling

### Backend

- [x] `ValueError` from validation is consistently caught and converted to `HTTPException(400)` or `HTTPException(403)` as appropriate
- [x] Custom component code endpoints return 403 (Forbidden) — correct HTTP semantics
- [x] MCP endpoint returns error text content instead of raising (matches MCP protocol)
- [x] OpenAI Responses endpoint returns `OpenAIErrorResponse` (matches OpenAI format)
- [x] No silent failures in backend
- [x] No empty catch blocks
- [x] No generic bare `except:` blocks

### Frontend

**Issue 1 — Silent `undefined` returns (MINOR)**

In `use-post-template-value.ts` (lines ~181/191), returning `undefined` silently skips the template value update when a component is outdated. The user receives no feedback that the operation was skipped.

**Issue 2 — 403 suppression could mask auth errors (MITIGATED)**

The 403 catch in `use-post-template-value.ts` and `use-refresh-model-inputs.ts` suppresses errors, but the check for `e.response.data.detail.includes("Custom component")` narrows it to only the specific custom component error. This is an acceptable mitigation.

**Issue 3 — Empty flow guard edge case (MINOR)**

In `use-save-flow.ts`, the condition:
```typescript
if (!currentSavedFlow?.data?.nodes.length || data!.nodes.length > 0)
```
Could prevent saving a legitimately empty flow (user intentionally deleted all nodes). The `!length` check is truthy when `length === 0`, so an empty flow would not be saved if the previous flow had nodes.

**Verdict: PASS (with minor frontend notes)**

---

## RECOMMENDED: Observability

- [x] `logger.warning` for blocked components with display names and node IDs
- [x] `logger.error` for fail-closed when hash cache unavailable
- [x] `logger.debug` for hash computation and component processing
- [x] Appropriate log levels (debug for tracing, warning for blocks, error for system issues)
- [x] **No sensitive data in logs**
- [x] No redundant logs

**Verdict: PASS**

---

## RECOMMENDED: Comments

- [x] Docstrings explain "why" — e.g., `flow_validation.py` module docstring explains the security model
- [x] `# SECURITY:` comments document trust boundaries (e.g., `chat.py` cache-hit path)
- [x] `# SYNC:` comments link related code across stack boundaries
- [x] No commented-out code
- [x] No TODO without context

**Verdict: PASS**

---

## TESTING

### Backend Tests

#### `test_flow_validation.py` (~487 lines, 4 test classes)

**Happy path tests:**
- [x] Current code hash passes validation
- [x] Custom components allowed when setting is True
- [x] Empty/None flow data passes

**Adversarial tests:**
- [x] `test_edited_false_with_custom_code_still_blocked` — core security property: client `edited=False` doesn't bypass server check
- [x] `test_security_edited_false_custom_code_blocked` — explicit security test with attacker scenario commentary
- [x] `test_fail_closed_when_hash_dict_is_none` — fail-closed behavior when server starting
- [x] `test_fail_closed_blocks_edited_true_without_cache` — double-checks fail-closed
- [x] Unknown types blocked
- [x] Hash mismatches detected as outdated
- [x] Nested nodes recursively checked

**Regression protection:**
- [x] `TestEndpointValidationCoverage` — uses `inspect.getsource()` to verify all execution endpoints contain `check_flow_and_raise` calls. This will catch future endpoints that forget validation.

**Edge cases:**
- [x] Nodes without code field
- [x] Empty component type
- [x] Missing metadata in component cache
- [x] Legacy alias resolution

#### `test_initial_setup.py` (~277 lines)

- [x] Hash_history stripping
- [x] Metadata preservation
- [x] Components without metadata
- [x] Legacy alias resolution
- [x] Direct key precedence over aliases

### Frontend Tests

#### `use-refresh-model-inputs.test.ts` (~180 lines)

- [x] Skipping outdated nodes when custom components disabled
- [x] Not skipping when custom components enabled
- [x] Not skipping user-edited nodes
- [x] Mixed outdated/fresh nodes
- [x] Template code mismatch fallback
- [x] 403 suppression fallback

#### `templatesGenerator.test.ts` (~111 lines)

- [x] Basic templating
- [x] Group node skipping
- [x] Legacy alias creation
- [x] Alias not overwriting direct keys
- [x] Missing targets
- [x] Graceful handling

### Test Quality Assessment

- [x] Clear test names: `test_should_block_when_...`, `test_should_allow_when_...`
- [x] Independent tests (no shared mutable state)
- [x] Deterministic (no time-dependent or external calls)
- [x] External dependencies mocked
- [x] Tests validate behavior, not implementation details
- [x] Clear AAA structure

### Missing Test Coverage

1. **No integration test** verifying the full endpoint-to-validation path (e.g., calling `/api/v1/build/{flow_id}/flow` with blocked components and verifying the 400 response). The unit tests mock the validation, but an integration test would catch wiring issues.
2. **No frontend test for `flowStore.buildFlow`** with the new `waitForNodeUpdates` + save + re-validate sequence.
3. **No test for the empty flow guard** in `use-save-flow.ts`.

### Coverage Validation

- [ ] **Coverage was NOT run and shown** — the PR does not include coverage reports for either backend or frontend tests. Per the testing rules, coverage should be run and >=75% on all tested source code.

**Verdict: PASS (with notes on missing coverage report and minor gaps)**

---

## LEGACY CODE AWARENESS

- [x] New code does NOT prolong bad patterns
- [x] Legacy alias handling (`"Prompt" -> "Prompt Template"`) is documented and isolated
- [x] New validation module is cleanly separated from existing code
- [x] `SYNC` comments document cross-stack dependencies

**Verdict: PASS**

---

## Pre-Submission Checklist

```
CRITICAL (Blockers)
[x] No PII in any logs, prints, or webhook messages
[x] No secrets/credentials in code
[~] No duplicate types, classes, or logic (DRY) — 3 violations identified, none blocking
[x] No file exceeds ~500 lines
[x] No file has more than 5 functions with DIFFERENT responsibilities
[x] No file has more than 10 functions even with same responsibility
[x] No file has more than 1 main class
[x] No mixed responsibility prefixes in same file

IMPORTANT (Must fix)
[x] Each file/function has single responsibility
[x] Proper error handling (no silent failures in backend)
[~] Strong typing — minor `any` usage in TS error handlers
[x] Inputs validated at boundaries
[x] Types in dedicated types file
[x] Constants in dedicated constants file

RECOMMENDED (Should fix)
[x] Appropriate logging at key points
[x] No unnecessary comments
[x] No over-engineering

TESTING
[x] Unit tests for core logic
[x] Tests cover success, error, and edge cases
[x] Tests have BOTH happy path AND adversarial tests
[x] Adversarial tests included: edge cases, invalid inputs, boundary values, error states
[ ] Coverage ran and shown — NOT PROVIDED
[x] All created tests appear well-structured
[x] Not prolonging legacy bad patterns
```

---

## Final Verdict

### Overall: APPROVE with non-blocking feedback

The PR implements a solid security feature with excellent design principles (fail-closed, server-side verification, client distrust). Test coverage is thorough with both happy path and adversarial tests, including clever regression protection via source inspection.

### Must address before merge:
- None (no blockers)

### Should address (non-blocking):

| # | Severity | Issue | Location |
|---|----------|-------|----------|
| 1 | MODERATE | DRY: Backend validation pattern repeated ~8 times | `chat.py`, `endpoints.py`, `mcp_utils.py`, etc. |
| 2 | SIGNIFICANT | DRY: Frontend guard logic duplicated in 2 hooks | `use-post-template-value.ts`, `use-refresh-model-inputs.ts` |
| 3 | MODERATE | DRY: Legacy aliases defined in 3 places | `setup.py`, `components.py`, `reactflowUtils.ts` |
| 4 | MINOR | `any` types in TypeScript error handlers | `deploy-dropdown.tsx`, `use-post-template-value.ts`, `use-save-flow.ts` |
| 5 | MINOR | Silent `undefined` return skips update without user feedback | `use-post-template-value.ts` |
| 6 | MINOR | Empty flow guard may prevent saving legitimately empty flows | `use-save-flow.ts` |
| 7 | INFO | Coverage report not included | Tests exist but >=75% not verified |
| 8 | INFO | No integration test for full endpoint-to-validation path | Test suite |

File: src/backend/base/langflow/api/v1/endpoints.py [hunk 1]
```diff
@@ -1009,6 +1018,19 @@ async def custom_component(
     raw_code: CustomComponentRequest,
     user: CurrentActiveUser,
 ) -> CustomComponentResponse:
+    settings_service = get_settings_service()
+    if not settings_service.settings.allow_custom_components:
+        # Allow updating to a known server template (core component update),
+        # but block truly custom code.
+        is_known_template = component_cache.all_known_hashes and code_hash_matches_any_template(
```
Comment (ogabrielluiz) [inline src/backend/base/langflow/api/v1/endpoints.py]: when `all_known_hashes` is `None` (cache still loading at startup), the short-circuit `and` makes `is_known_template` falsy, and the 403 fires with "Custom component creation is disabled." this is fail-closed which is great for security, but unlike `validate_flow_for_current_settings` which calls `_get_component_hash_lookups_for_validation()` to lazily compute hashes from `all_types_dict`, this path doesn't attempt that lazy computation.

so during startup, legitimate component updates that match a known template get a misleading "disabled" error instead of either succeeding or getting a "still initializing" message. might be worth using the same lazy-compute pattern here, or at least distinguishing the "cache not ready" case from the "truly disabled" case in the error message.

(same pattern appears in `custom_component_update` below)

File: src/backend/base/langflow/api/v1/chat.py [hunk 2]
```diff
@@ -648,6 +671,8 @@ async def build_public_tmp(
             queue_service=queue_service,
             flow_name=flow_name or f"{client_id}_{flow_id}",
```
Comment (ogabrielluiz) [inline src/backend/base/langflow/api/v1/chat.py:714]: since `CustomComponentValidationError` extends `ValueError`, this catch will intercept it and pass `str(exc)` directly as the HTTP detail. the error message includes internal node display names and IDs (e.g., `"Flow build blocked: custom components are not allowed: SomeNode (node-abc123)"`). for an unauthenticated public endpoint, this leaks internal flow structure to anonymous callers.

might be worth returning a generic message here like `"This flow cannot be executed"` and logging the detailed message server-side instead.

File: src/lfx/src/lfx/load/load.py [hunk 3]
```diff
@@ -79,6 +80,8 @@ async def aload_flow_from_json(
     if tweaks is not None:
         graph_data = process_tweaks(graph_data, tweaks)
 
+    await ensure_component_hash_lookups_loaded()
+
     from lfx.graph.graph.base import Graph
```
Comment (ogabrielluiz) [inline src/lfx/src/lfx/load/load.py:94]: if this call fails (e.g., `get_and_cache_all_types_dict` throws), the underlying exception re-raises as-is (could be any type). but `Graph.from_payload` below would raise `CustomComponentValidationError` for the same conceptual problem (cache not ready). callers of `aload_flow_from_json` that only catch `CustomComponentValidationError` won't handle the raw exception from this path.

might be worth wrapping the failure as `CustomComponentValidationError` or `RuntimeError` with a clear message so callers have a consistent error contract.

File: src/lfx/src/lfx/utils/flow_validation.py [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+    dump_graph = getattr(graph, "dump", None)
+    if callable(dump_graph):
+        dumped_graph = dump_graph()
+        if isinstance(dumped_graph, Mapping):
+            dumped_data = dumped_graph.get("data")
+            if isinstance(dumped_data, Mapping):
+                return dumped_data
+
+    return raw_graph_data
```
Comment (ogabrielluiz) [inline src/lfx/src/lfx/utils/flow_validation.py]: I think this may be a dangerous assumption. Could be wrong, though, but I'm not sure `dump` always has what it needs to function properly.

File: src/lfx/src/lfx/graph/graph/base.py [hunk 5]
```diff
@@ -1163,8 +1163,17 @@ def from_payload(
...
+
         if "data" in payload:
             payload = payload["data"]
+        # Defense-in-depth: validate here so that no code path can construct
+        # a graph with blocked/custom components, even if an API endpoint
+        # forgets its own pre-check. Ideally this would live only at the API
+        # boundary (middleware or endpoint dependency) so from_payload stays
+        # pure deserialization, but a missed endpoint means arbitrary code
+        # execution, so we keep this as a safety net.
+        validate_flow_for_current_settings(payload)
```
Comment (ogabrielluiz) [inline src/lfx/src/lfx/graph/graph/base.py:1178]: Good call

**Comment by tedgoddard** [comment]
This feature is much appreciated and is working well for me. I tested with "custom components" added via LANGFLOW_COMPONENTS_PATH and they continued to work as expected, however new custom components could not be added via the UI when LANGFLOW_ALLOW_CUSTOM_COMPONENTS=False was set.

**Comment by jordanrfrazier** [comment]
> This feature is much appreciated and is working well for me. I tested with "custom components" added via LANGFLOW_COMPONENTS_PATH and they continued to work as expected, however new custom components could not be added via the UI when LANGFLOW_ALLOW_CUSTOM_COMPONENTS=False was set.

@tedgoddard Glad to hear it's working well. We'll aim to get this merged and released with our upcoming v1.9.0 release, and soon to come with additional features, like allowing admins to specify a list of allowed custom components on setup. If you have any other requests or ideas, please share and I'll see what I can do.
